### DISCUSSION TITLE
I had try to read FrameGraph-Example

### DISCUSSION THREAD

**Discussion by FuXiii** [skaarj1989/FrameGraph#3] [General]
Hello, it's me again~
The FrameGraph-Example project was amazing. the `PBR`, `IBL` and even the `Global Illumination`. I am just a fresh hand in Computer Graph. How did you learn this knowledge about the Computer Graph? I also want learn them but I don't know where to start it. I just know how to use `Vulkan` to render a 3D model and skybox. such like `vkRenderPass`,`Subpass` and `vkGraphicsPipeline` etc, But I don't know how to put them together in to FrameGraph. Such like how to design the `Material`  and `Mesh` class and combine `Vulkan` with `FrameGraph`, can you give me some advice or recommended learning path.

**Comment by skaarj1989** [comment]
I have been playing with graphics programming for 2 years (last December I started to use Vulkan).
I made tons of mistakes (I still do) and rewritten my "game engine" countless times.

You have to understand (and accept) the fact that you can not write/design the perfect (or even good) solution on your first attempt.

Just do not give up, code, experiment and enjoy it.

I can recommend you the following discord servers:
- [GraphicsProgramming](https://discord.gg/6mgNGk7)
- [Vulkan](https://discord.gg/vulkan)

**Comment by FuXiii** [comment]
Thank you~
I had joined into [GraphicsProgramming](https://discord.gg/6mgNGk7) discussion group and the [Vulkan](https://github.com/skaarj1989/FrameGraph/discussions/discord.gg/vulkan) url link I can not open it.
I had found the document about [GraphicsProgramming](https://discord.gg/6mgNGk7) in https://graphics-programming.org/

---

In FrameGraph-Example project, When I read `void GBufferPass::addGeometryPass(...)` function in `GBufferPass.cpp`:
```CXX
for (const auto &renderable : renderables) {
        auto &[mesh, subMeshIndex, material, flags, modelMatrix, _] =
          *renderable;

        rc.setGraphicsPipeline(_getPipeline(*mesh.vertexFormat, &material))
          .bindUniformBuffer(0, getBuffer(resources, frameBlock));

        _setTransform(*camera, modelMatrix);
        for (uint32_t unit{kFirstFreeTextureBinding};
             const auto &[_, texture] : material.getDefaultTextures()) {
          rc.bindTexture(unit++, *texture);
        }
        rc.setUniform1i("u_MaterialFlags", flags)
          .draw(*mesh.vertexBuffer, *mesh.indexBuffer,
                mesh.subMeshes[subMeshIndex].geometryInfo);
      }
```
It will draw all the visible and opaque renderable object.

Is that means all the renderable object will rendering in `FrameGraph::PassNode::Execute()`?

---

If all the rendering will in `FrameGraph::PassNode::Execute()`, If in sence there have two object one is cube and another is sphere. just render them. and the FrameGraph is simple:
![image](https://user-images.githubusercontent.com/40132375/184264897-d2ab10a6-5be7-47d9-b098-d0d280fb12e7.png)
in `SimpleDrawPass::Execute()` can I write it like this?:
```CXX
SimpleDrawPass:::Execute()
{
        vkFrameBuffer fb = {renderTarget, depthImage};
        vkCmdBeginRenderPass(fb,...);
        ...
        vkCmdDraw(cube);
        ...
        vkCmdDraw(sphere);
        vkCmdEndRenderPass()
        
        CmdBufferSubmitToQueue(cmdBuffer);   
}
```
Is that means a `PassNode` in FrameGraph is an `vkRenderPass`?

If the FrameGraph like this:
![image](https://user-images.githubusercontent.com/40132375/184265324-cf792c93-7207-4f07-bb81-f794473c9db2.png)
the `SimpleDrawPass::Execute()` similar as above. 
```CXX
SimpleDrawPass:::Execute()
{
        vkFrameBuffer fb = {colorImage, depthImage};
        vkCmdBeginRenderPass(fb,...);
        ...
        vkCmdDraw(cube);
        ...
        vkCmdDraw(sphere);
        vkCmdEndRenderPass()
        
        CmdBufferSubmitToQueue(cmdBuffer);   
}
```
but the `SamplePass::Execute()` maybe looks like:
```CXX
SamplePass:::Execute()
{
        vkFrameBuffer fb = {colorImage, depthImage, renderTarget};
        vkCmdBeginRenderPass(fb,...);
        ...
        vkCmdDraw(wholeSurface);//sample from colorImage or depthImage into renderTarget, such like use input attachment
        vkCmdEndRenderPass()
        
        CmdBufferSubmitToQueue(cmdBuffer);   
}
```
In this case we did't use `Subpass` feature. But normally, use `Subpass` these can be combined into one commandBuffer such like:
```
vkFrameBuffer fb = {colorImage, depthImage, renderTarget};
vkCmdBeginRenderPass(fb,...);
...
vkCmdDraw(cube);
...
vkCmdDraw(sphere);

vkCmdNextSubpass();
...
vkCmdDraw(wholeSurface);//sample from colorImage or depthImage into renderTarget, such like use input attachment
vkCmdEndRenderPass()
        
CmdBufferSubmitToQueue(cmdBuffer);   
```
this means a `FrameGraph` become a `RenderPass`,and `PassNode` become `Subpass`. Is that right?How to elegent combine `PassNode`, `RenderPass` and `Subpass`?

---
Can you share how to learn `Global Illumination`? Thanks~

**Comment by skaarj1989** [comment]
In my Vulkan RHI I use `VK_KHR_dynamic_rendering` (which means I do not need `VkRenderPass` and `VkFramebuffer`).

Global Illumination overview: https://morgan3d.github.io/articles/2019-04-01-ddgi/overview.html

**Comment by FuXiii** [comment]
Yes, if the device support vulkan 1.3 the `VK_KHR_dynamic_rendering` will be fine, but in some vender's physical device they only support vulkan 1.0.  o(╥﹏╥)o~

**Comment by FuXiii** [comment]
in [Global Illumination overview](https://morgan3d.github.io/articles/2019-04-01-ddgi/overview.html). Is this GI implemented using ray tracing？such like use `VK_NV_ray_tracing` in vulkan by NVIDIA GPU card or `VK_KHR_ray_tracing_pipeline`?
